File: java/src/Ui.java
// Ui.java - the pieces every lesson uses to draw the screen.
//
// Same purpose and the same visual style as the other courses' ui modules,
// kept as a separate implementation rather than shared code -- see
// docs/writing-a-course.md for why. Static methods on a final class rather
// than free functions, because that is what Java has; IN is a single
// Scanner opened once over System.in and reused everywhere, the same way
// the C++ course reuses a single std::cin.

import java.util.List;
import java.util.Scanner;

public final class Ui {
    private static final int WIDTH = 54;

    // Package-private, not private: lessons that need a raw read (Scanner's
    // nextInt()/nextLine() themselves are what module 3 teaches) share this
    // one Scanner rather than each opening their own over System.in, which
    // risks two Scanners racing over the same underlying stream.
    static final Scanner IN = new Scanner(System.in);

    private Ui() {
    }

    public static void clearScreen() {
        // ANSI codes, not ProcessBuilder("clear").inheritIO().start(): no
        // dependency on TERM being set. Same fix, same reason, as every
        // other course's ui module.
        System.out.print("\033[H\033[2J\033[3J");
        System.out.flush();
    }

    public static void waitEnter() {
        System.out.print("\n  Press ENTER to continue...");
        System.out.flush();
        IN.nextLine();
    }

    public static void rule() {
        System.out.println("  " + "-".repeat(WIDTH));
    }

    private static void paddedLine(String text, char border) {
        String padded = text.length() < WIDTH - 3
                ? text + " ".repeat(WIDTH - 3 - text.length())
                : text;
        System.out.println("  " + border + " " + padded + border);
    }

    private static void frame(char fill) {
        System.out.println("  +" + String.valueOf(fill).repeat(WIDTH - 2) + "+");
    }

    public static void title(String text) {
        System.out.println();
        frame('=');
        paddedLine(text, '|');
        frame('=');
        System.out.println();
    }

    public static void heading(String text) {
        System.out.println("\n  " + text);
        rule();
    }

    public static boolean askYes(String questionText) {
        System.out.print("\n  " + questionText + " (y/N): ");
        System.out.flush();
        String answer = IN.nextLine();
        return !answer.isEmpty() && (answer.charAt(0) == 'y' || answer.charAt(0) == 'Y');
    }

    public static void exercise(int number) {
        System.out.println("\n  >> EXERCISE - MODULE " + number);
        rule();
    }

    public static boolean question(String text, String correct, String why) {
        System.out.println("\n  " + text);
        System.out.print("  Your answer: ");
        System.out.flush();
        String answer = IN.nextLine();

        boolean right = answer.strip().equalsIgnoreCase(correct.strip());
        if (right) {
            System.out.println("\n  CORRECT.  " + why);
        } else {
            System.out.println("\n  NOT QUITE. The answer is: " + correct);
            System.out.println("             " + why);
        }
        return right;
    }

    // A task to write in a real file.
    //
    // `input` is every line the program will read while producing
    // `expected` -- empty for a task that reads nothing. Together they are
    // the actual specification: run the solution, type `input`, get
    // `expected`, however the code that does it is written. `solution` is
    // one way of getting there and appears only after a confirmation.
    public static void challenge(List<String> task, List<String> input,
            List<String> expected, List<String> solution) {
        System.out.println("\n  >> WRITE THIS YOURSELF, in a real file");
        rule();

        for (String line : task) {
            System.out.println("  " + line);
        }

        if (!input.isEmpty()) {
            System.out.println("\n  Try it with this input:\n");
            for (String line : input) {
                System.out.println("      " + line);
            }
        }

        if (!expected.isEmpty()) {
            System.out.println("\n  It must print:\n");
            for (String line : expected) {
                System.out.println("      " + line);
            }
            System.out.println("\n  That output is the whole specification. Any code that");
            System.out.println("  produces it is correct.");
        }

        System.out.println("\n  Try it first. Compile with:");
        System.out.println("    javac Test.java && java Test");

        if (!askYes("Want to see example code?")) {
            return;
        }

        System.out.println();
        rule();
        for (String line : solution) {
            System.out.println("  " + line);
        }
        rule();
        System.out.println("  This is EXAMPLE CODE, not the answer. It is one way to get");
        System.out.println("  that output; yours may look nothing like it and still be");
        System.out.println("  right -- or better. Compare the output, not the code.");
    }
}

File: csharp/src/Ui.cs
// Ui.cs - the pieces every lesson uses to draw the screen.
//
// Same purpose and the same visual style as the other courses' ui modules,
// kept as a separate implementation rather than shared code -- see
// docs/writing-a-course.md for why. A static class, same shape as Java's
// Ui, since C# has the identical "no free functions" constraint Java does.

static class Ui
{
    const int Width = 54;

    public static void ClearScreen()
    {
        // ANSI codes, not Process.Start("clear"): no dependency on TERM
        // being set. Same fix, same reason, as every other course's ui
        // module.
        Console.Write("\x1b[H\x1b[2J\x1b[3J");
    }

    public static void WaitEnter()
    {
        Console.Write("\n  Press ENTER to continue...");
        Console.ReadLine();
    }

    public static void Rule() => Console.WriteLine("  " + new string('-', Width));

    static void PaddedLine(string text, char border)
    {
        string padded = text.Length < Width - 3
            ? text + new string(' ', Width - 3 - text.Length)
            : text;
        Console.WriteLine($"  {border} {padded}{border}");
    }

    static void Frame(char fill) => Console.WriteLine("  +" + new string(fill, Width - 2) + "+");

    public static void Title(string text)
    {
        Console.WriteLine();
        Frame('=');
        PaddedLine(text, '|');
        Frame('=');
        Console.WriteLine();
    }

    public static void Heading(string text)
    {
        Console.WriteLine($"\n  {text}");
        Rule();
    }

    public static bool AskYes(string questionText)
    {
        Console.Write($"\n  {questionText} (y/N): ");
        string? answer = Console.ReadLine();
        return !string.IsNullOrEmpty(answer) && (answer[0] == 'y' || answer[0] == 'Y');
    }

    public static void Exercise(int number)
    {
        Console.WriteLine($"\n  >> EXERCISE - MODULE {number}");
        Rule();
    }

    public static bool Question(string text, string correct, string why)
    {
        Console.WriteLine($"\n  {text}");
        Console.Write("  Your answer: ");
        string? answer = Console.ReadLine();

        bool right = string.Equals(answer?.Trim(), correct.Trim(), StringComparison.OrdinalIgnoreCase);
        if (right)
        {
            Console.WriteLine($"\n  CORRECT.  {why}");
        }
        else
        {
            Console.WriteLine($"\n  NOT QUITE. The answer is: {correct}");
            Console.WriteLine($"             {why}");
        }
        return right;
    }

    // A task to write in a real file.
    //
    // `input` is every line the program will read while producing
    // `expected` -- empty for a task that reads nothing. Together they are
    // the actual specification: run the solution, type `input`, get
    // `expected`, however the code that does it is written. `solution` is
    // one way of getting there and appears only after a confirmation.
    public static void Challenge(string[] task, string[] input, string[] expected, string[] solution)
    {
        Console.WriteLine("\n  >> WRITE THIS YOURSELF, in a real file");
        Rule();

        foreach (string line in task)
        {
            Console.WriteLine($"  {line}");
        }

        if (input.Length > 0)
        {
            Console.WriteLine("\n  Try it with this input:\n");
            foreach (string line in input)
            {
                Console.WriteLine($"      {line}");
            }
        }

        if (expected.Length > 0)
        {
            Console.WriteLine("\n  It must print:\n");
            foreach (string line in expected)
            {
                Console.WriteLine($"      {line}");
            }
            Console.WriteLine("\n  That output is the whole specification. Any code that");
            Console.WriteLine("  produces it is correct.");
        }

        Console.WriteLine("\n  Try it first. Run with:");
        Console.WriteLine("    dotnet run");

        if (!AskYes("Want to see example code?"))
        {
            return;
        }

        Console.WriteLine();
        Rule();
        foreach (string line in solution)
        {
            Console.WriteLine($"  {line}");
        }
        Rule();
        Console.WriteLine("  This is EXAMPLE CODE, not the answer. It is one way to get");
        Console.WriteLine("  that output; yours may look nothing like it and still be");
        Console.WriteLine("  right -- or better. Compare the output, not the code.");
    }
}
